drivers/serial/uart_rpi_pico.c

```diff
@@ -40,36 +51,249 @@ static void uart_rpi_poll_out(const struct device *dev, unsigned char c)
 static int uart_rpi_init(const struct device *dev)
 {
 	const struct uart_rpi_config *config = dev->config;
-	int ret;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	struct uart_rpi_data * const data = dev->data;
+	int ret, baudrate;
 
 	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
 	if (ret < 0) {
 		return ret;
 	}
 
-	uart_init(config->uart_dev, config->baudrate);
+	baudrate = uart_init(uart_inst, data->baudrate);
+	/* Check if baudrate adjustment returned by 'uart_init' function is a positive value */
+	if (baudrate <= 0) {
+		return -EINVAL;
+	}
+
+	hw_clear_bits(&uart_get_hw(uart_inst)->lcr_h, UART_UARTLCR_H_FEN_BITS);
+	uart_get_hw(uart_inst)->dr = 0U;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	config->irq_config_func(dev);
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
 	return 0;
 }
 
+static int uart_rpi_err_check(const struct device *dev)
+{
+	const struct uart_rpi_config * const config = dev->config;
+	uint32_t data_reg = uart_get_hw(config->uart_dev)->dr;
+	int errors = 0;
+
+	if (data_reg & UART_UARTDR_OE_BITS) {
+		errors |= UART_ERROR_OVERRUN;
+	}
+
+	if (data_reg & UART_UARTDR_BE_BITS) {
+		errors |= UART_BREAK;
+	}
+
+	if (data_reg & UART_UARTDR_PE_BITS) {
+		errors |= UART_ERROR_PARITY;
+	}
+
+	if (data_reg & UART_UARTDR_FE_BITS) {
+		errors |= UART_ERROR_FRAMING;
+	}
+
+	return errors;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static int uart_rpi_fifo_fill(const struct device *dev,
+			      const uint8_t *tx_data, int len)
+{
+	const struct uart_rpi_config * const config = dev->config;
+	uart_inst_t * const uart_inst = config->uart_dev;
+	int tx_len = 0;
+
+	while (uart_is_writable(uart_inst) && (len - tx_len) > 0) {
```

Comment:
this driver continues to mix API and register access. I think it would be better to just use register access everywhere, so for example:

```suggestion
	while (!(uart_get_hw(uart)->fr & UART_UARTFR_TXFF_BITS) && (len - tx_len) > 0) {
```

`uart_get_hw` could actually be removed by just storing a `uart_hw_t *`.

Reply:
Okay, I understand now, I will do this modifications soon.

Reply:
@gmarull What about uart_init function?

Reply:
note: `uart_init()` actions should also be handled by the driver

Reply:
Note that uart_init function is using reset and clock API functions, but reset driver and clock driver are not implemented in Zephyr yet. If I want to implement the init function I need to bring another API functions. I think for now this API call is not dangerous and in the future I would like to implement those drivers in Zephyr and then I will modify also this function.

Reply:
Fair enough, I guess it's ok for now then.

Reply:
Would you mind adding a comment around `uart_init()`?

Reply:
A comment like?

Reply:
```
/* uart_init() may be replaced by register based API once rpi-pico platform has a clock controller driver */
```

Reply:
Done! Thanks for review. :) Tomorrow I will make a PR with I2C driver.